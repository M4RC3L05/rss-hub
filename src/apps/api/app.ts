import Koa from "koa";
import bodyParser from "koa-bodyparser";
import cors from "@koa/cors";
import qs from "koa-qs";
import basicAuth from "koa-basic-auth";
import config from "config";
import { errorMapper, requestLifeCycle } from "../../middlewares/mod.js";
import { errorMappers } from "../../errors/mod.js";
import { router } from "./router.js";

const makeApp = () => {
  const app = new Koa();

  qs(app);

  app.use(requestLifeCycle);
  app.use(
    errorMapper({
      mappers: [errorMappers.validationErrorMapper, errorMappers.sqliteErrorMapper],
      defaultMapper: errorMappers.defaultErrorMapper,
    }),
  );
  app.use(cors());
  app.use(basicAuth({ ...config.get<{ name: string; pass: string }>("apps.api.basicAuth") }));
  app.use(bodyParser());
  app.use(router.routes());
  app.use(router.allowedMethods());

  return app;
};

export default makeApp;
