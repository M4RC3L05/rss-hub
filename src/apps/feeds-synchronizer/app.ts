import { sql } from "@m4rc3l05/sqlite-tag";
import { formatError, makeLogger } from "#src/common/logger/mod.ts";
import type { CustomDatabase } from "#src/database/mod.ts";
import type { FeedsTable } from "#src/database/types/mod.ts";
import type { FeedService } from "#src/services/mod.ts";

const log = makeLogger("feed-synchronizer-runner");

const runner = async ({
  signal,
  feedService,
  db,
}: { signal?: AbortSignal; db: CustomDatabase; feedService: FeedService }) => {
  const feeds = db.all<FeedsTable>(sql`select * from feeds`);

  log.info("Synching begin");

  for (const feed of feeds) {
    log.info(`Synching feed ${feed.url}`, { feed });

    try {
      const { faildCount, failedReasons, successCount, totalCount } =
        await feedService.syncFeed(feed, { signal });

      log.info(
        `Done processing ${feed.url}, ${totalCount} items, with ${successCount} succeeded and ${faildCount} failed`,
        {
          failedReasons: failedReasons.map((reason) =>
            reason instanceof Error ? formatError(reason) : reason
          ),
          faildCount,
          successCount,
          totalCount,
          feed,
        },
      );
    } catch (error) {
      log.error(`Sync failed ${feed.url}`, { error });
    } finally {
      log.info(`Synching synced ${feed.url}`, { feed });
    }

    if (signal?.aborted) {
      break;
    }
  }

  log.info("Synching ended");
};

export default runner;
