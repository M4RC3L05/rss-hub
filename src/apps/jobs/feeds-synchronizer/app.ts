import pineSerializer from "pino-std-serializers";
import { makeLogger } from "#src/common/logger/mod.ts";
import type { CustomDatabase, FeedsTable } from "#src/database/mod.ts";
import type { FeedService } from "#src/services/mod.ts";

const log = makeLogger("feed-synchronizer-runner");

const runner = async ({
  signal,
  feedService,
  db,
}: { signal?: AbortSignal; db: CustomDatabase; feedService: FeedService }) => {
  if (signal?.aborted) {
    return;
  }

  const feeds = db.sql<FeedsTable>`select * from feeds`;

  log.info("Synching begin");

  for (const feed of feeds) {
    if (signal?.aborted) {
      log.info("Abort sync");

      break;
    }

    log.info(`Synching feed ${feed.url}`, { feed });

    try {
      const { faildCount, failedReasons, successCount, totalCount } =
        await feedService.syncFeed(feed, signal ? { signal } : undefined);

      log.info(
        `Done processing ${feed.url}, ${totalCount} items, with ${successCount} succeeded and ${faildCount} failed`,
        {
          failedReasons: failedReasons.map((reason: unknown) =>
            reason instanceof Error
              ? pineSerializer.errWithCause(reason)
              : reason
          ),
          faildCount,
          successCount,
          totalCount,
          feed,
        },
      );
    } catch (error) {
      log.error(`Sync failed ${feed.url}`, { error });
    } finally {
      log.info(`Synching synced ${feed.url}`, { feed });
    }

    if (signal?.aborted) {
      break;
    }
  }

  log.info("Synching ended");
};

export default runner;
