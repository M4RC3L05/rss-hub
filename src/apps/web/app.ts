import { serveStatic } from "@hono/node-server/serve-static";
import config from "config";
import { type ContextVariableMap, Hono } from "hono";
import { basicAuth } from "hono/basic-auth";
import { HTTPException } from "hono/http-exception";
import { secureHeaders } from "hono/secure-headers";
import { makeLogger } from "#src/common/logger/mod.js";
import type { DeepPartial } from "#src/common/utils/types.js";
import { requestLifeCycle, serviceRegister } from "#src/middlewares/mod.js";
import { router } from "./router.js";

const log = makeLogger("web");

const makeApp = (deps: DeepPartial<ContextVariableMap>) => {
  const app = new Hono();

  app.use("*", requestLifeCycle);
  app.use("*", serviceRegister(deps));
  app.use("*", secureHeaders());
  app.use("*", async (c, next) => {
    try {
      await next();
    } finally {
      // This is important so that we always make sure the browser will not cache the previous page
      // so that the requests are always made.
      if (
        !c.req.path.startsWith("/public") &&
        !c.req.path.startsWith("/deps")
      ) {
        c.header("cache-control", "no-cache, no-store, must-revalidate");
        c.header("pragma", "no-cache");
        c.header("expires", "0");
      }
    }
  });
  app.use(
    "*",
    basicAuth({
      username: config.get<{ name: string; pass: string }>("apps.web.basicAuth")
        .name,
      password: config.get<{ name: string; pass: string }>("apps.web.basicAuth")
        .pass,
    }),
  );

  app.get("/favicon.ico", serveStatic({ root: "./src/apps/web/public" }));
  app.get(
    "/public/*",
    serveStatic({
      root: "./src/apps/web/public",
      rewriteRequestPath: (path) => path.replace("/public", ""),
    }),
  );
  app.route(
    "/deps",
    new Hono()
      .get(
        "/simpledotcss/*",
        serveStatic({
          root: "./node_modules/simpledotcss",
          rewriteRequestPath: (path) => path.replace("/deps/simpledotcss", ""),
        }),
      )
      .get(
        "/htmx.org/*",
        serveStatic({
          root: "./node_modules/htmx.org",
          rewriteRequestPath: (path) => path.replace("/deps/htmx.org", ""),
        }),
      ),
  );

  router(app);

  app.onError((error, c) => {
    log.error(
      error instanceof Error || typeof error === "object" ? error : { error },
      "Something went wrong!",
    );

    if (error instanceof HTTPException) {
      if (error.res) {
        for (const [key, value] of error.res.headers.entries()) {
          c.header(key, value);
        }
      }
    }

    return c.text(
      error.message ? error.message : "Something broke",
      (error as Error & { status: number }).status ?? 500,
    );
  });

  app.notFound(() => {
    throw new HTTPException(404, { message: "Route not found" });
  });

  return app;
};

export default makeApp;
