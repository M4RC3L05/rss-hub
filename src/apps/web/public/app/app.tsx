import Masonry from "masonry-layout";
import { type FC, useCallback, useEffect, useMemo, useRef } from "react";
import { Col, Container, Row } from "react-bootstrap";
import useSWR from "swr";
import {
  type CategoriesTable,
  type FeedsTable,
} from "../../../../database/types/mod.js";
import { paths } from "./common/api.js";
import {
  CategoryItem,
  CreateCategoryItem,
} from "./components/categories/mod.js";

const App: FC = () => {
  const { data: categories } = useSWR<
    Array<CategoriesTable & { feedCount: number }>
  >(paths.categories.getCategories);
  const categoryIds = useMemo(
    () => categories?.map(({ id }) => id) ?? [],
    [categories],
  );
  const {
    data: feeds,
    isLoading,
    isValidating,
  } = useSWR<
    Record<
      string,
      Array<FeedsTable & { unreadCount: number; bookmarkedCount: number }>
    >
  >(
    categoryIds.length > 0
      ? `${paths.feeds.getFeeds}?${categoryIds.reduce(
          (acc, cid, index) =>
            `${acc}${index === 0 ? "" : "&"}categoryId=${cid}`,
          "",
        )}`
      : null,
    { refreshInterval: 10_000 },
  );
  const rowRef = useRef<HTMLDivElement>();
  const masonryRef = useRef<Masonry>();

  useEffect(() => {
    if (!masonryRef.current && rowRef.current) {
      masonryRef.current = new Masonry(rowRef.current, {
        percentPosition: true,
        initLayout: false,
        itemSelector: ".mesonry-item",
      });
    }
  }, []);

  useEffect(() => {
    if (categories && categories?.length >= 0) {
      masonryRef.current?.reloadItems?.();
      masonryRef.current?.layout?.();
    }
  }, [categories]);

  useEffect(() => {
    if (!isLoading) {
      masonryRef.current?.reloadItems?.();
      masonryRef.current?.layout?.();
    }
  }, [isLoading]);

  useEffect(() => {
    if (!isValidating) {
      masonryRef.current?.reloadItems?.();
      masonryRef.current?.layout?.();
    }
  }, [isValidating]);

  const renderCategoryItem = useCallback(
    ({
      category,
      feeds = [],
    }: {
      category: CategoriesTable & { feedCount: number };
      feeds: Array<
        FeedsTable & { unreadCount: number; bookmarkedCount: number }
      >;
    }) => <CategoryItem category={category} feeds={feeds} key={category.id} />,
    [],
  );

  return (
    <Container className="py-4">
      <Row className="mb-4">
        <Col className="text-center">
          <h1 className="display-2">RSS hub</h1>
          <h4>Track and get the latests new</h4>
        </Col>
      </Row>
      <Row ref={rowRef}>
        <CreateCategoryItem />
        {(categories ?? []).map((category) =>
          renderCategoryItem({ category, feeds: feeds?.[category.id] ?? [] }),
        )}
      </Row>
    </Container>
  );
};

export default App;
