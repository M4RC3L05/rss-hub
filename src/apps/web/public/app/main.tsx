import { SWRConfig } from "swr";
import { StrictMode } from "react";
// eslint-disable-next-line n/file-extension-in-import
import { createRoot } from "react-dom/client";
import { makeRequester } from "./common/api.js";
import App from "./app.js";

const Root = () => {
  return (
    <SWRConfig value={{ fetcher: makeRequester }}>
      <App />
    </SWRConfig>
  );
};

createRoot(document.querySelector("#app")!).render(
  <StrictMode>
    <Root />
  </StrictMode>,
);
