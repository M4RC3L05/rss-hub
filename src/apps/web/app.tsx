import config from "config";
import { type ContextVariableMap, Hono } from "@hono/hono";
import { serveStatic } from "@hono/hono/deno";
import { basicAuth } from "@hono/hono/basic-auth";
import { jsxRenderer } from "@hono/hono/jsx-renderer";
import { HTTPException } from "@hono/hono/http-exception";
import { secureHeaders } from "@hono/hono/secure-headers";
import { makeLogger } from "#src/common/logger/mod.ts";
import { serviceRegister } from "#src/middlewares/mod.ts";
import { router } from "#src/apps/web/routes/mod.ts";
import type {
  CategoriesService,
  FeedItemsService,
  FeedsService,
  OpmlService,
} from "#src/apps/web/services/api/mod.ts";
import { MainLayout } from "#src/apps/web/views/common/layouts/main.tsx";

const log = makeLogger("web");

declare module "@hono/hono" {
  interface ContextVariableMap {
    services: {
      api: {
        categoriesService: CategoriesService;
        opmlService: OpmlService;
        feedItemsService: FeedItemsService;
        feedsService: FeedsService;
      };
    };
  }
}

export const makeApp = (deps: Partial<ContextVariableMap>) => {
  const app = new Hono();

  app.onError((error, c) => {
    log.error("Something went wrong!", { error });

    if (error instanceof HTTPException) {
      if (error.res) {
        error.res.headers.forEach((value, key) => {
          c.header(key, value);
        });
      }
    }

    // Redirect back on request that alter the application state.
    if (!["GET", "HEAD", "OPTIONS"].includes(c.req.method)) {
      return c.redirect(c.req.header("Referer") ?? "/");
    }

    return c.text(
      error.message ?? "Something broke",
      // deno-lint-ignore no-explicit-any
      (error as any).status ?? 500,
    );
  });

  app.notFound(() => {
    throw new HTTPException(404, { message: "Route not found" });
  });

  app.use("*", serviceRegister(deps));
  app.use("*", secureHeaders({ referrerPolicy: "same-origin" }));
  app.use("*", async (c, next) => {
    try {
      await next();
      // This is important so that we always make sure the browser will not cache the previous page
      // so that the requests are always made.
      c.header("cache-control", "no-cache, no-store, must-revalidate");
      c.header("pragma", "no-cache");
      c.header("expires", "0");
    } catch (error) {
      // This is important so that we always make sure the browser will not cache the previous page
      // so that the requests are always made.
      c.header("cache-control", "no-cache, no-store, must-revalidate");
      c.header("pragma", "no-cache");
      c.header("expires", "0");

      throw error;
    }
  });
  app.use(
    "*",
    basicAuth({
      ...config.get<{ name: string; pass: string }>("apps.web.basicAuth"),
    }),
  );

  app.get("/favicon.ico", serveStatic({ root: "./src/apps/web/public" }));
  app.get(
    "/public/*",
    serveStatic({
      root: "./src/apps/web/public",
      rewriteRequestPath: (path) => path.replace("/public", ""),
    }),
  );

  app.get(
    "*",
    jsxRenderer(({ children }) => <MainLayout>{children}</MainLayout>, {
      docType: true,
      stream: true,
    }),
  );

  return app.route("/", router());
};
