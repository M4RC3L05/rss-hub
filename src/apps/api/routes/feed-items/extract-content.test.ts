import { type CustomDatabase, makeDatabase } from "#src/database/mod.ts";
import { makeApp } from "#src/apps/api/app.ts";
import {
  afterAll,
  beforeAll,
  beforeEach,
  describe,
  it,
} from "@std/testing/bdd";
import { assertSpyCallArg, assertSpyCalls, spy, stub } from "@std/testing/mock";
import { assertSnapshot } from "@std/testing/snapshot";
import { encodeBase64 } from "@std/encoding";
import { assertEquals } from "@std/assert";
import type { Hono } from "@hono/hono";
import { testDbUtils } from "#src/common/utils/mod.ts";
import * as testFixtures from "#src/common/test-fixtures/mod.ts";

let db: CustomDatabase;
let app: Hono;

beforeAll(async () => {
  db = makeDatabase();
  await testDbUtils.runMigrations(db);

  app = makeApp({ database: db, shutdown: new AbortController().signal });
});

beforeEach(() => {
  db.exec("delete from categories");
  db.exec("delete from feeds");
  db.exec("delete from feed_items");
});

afterAll(() => {
  db.close();
});

describe("GET /api/feed-items/:id/:feedId/extract-content", () => {
  it("should throw a validation error if invalid request params data is provided", async () => {
    const response = await app.request(
      "/api/feed-items/foo/bar/extract-content",
      {
        headers: {
          "authorization": `Basic ${encodeBase64("foo:bar")}`,
        },
      },
    );

    const data = await response.json();

    assertEquals(response.status, 422);
    assertEquals(data, {
      error: {
        code: "error",
        message: "Validation failed",
        validationErrors: [
          {
            field: "feedId",
            message: "The feedId field must be a valid UUID",
            rule: "uuid",
          },
        ],
      },
    });
  });

  it("should throw a validation error if invalid request params data is provided", async () => {
    const response = await app.request(
      `/api/feed-items/foo/${crypto.randomUUID()}/extract-content`,
      {
        headers: {
          "authorization": `Basic ${encodeBase64("foo:bar")}`,
        },
      },
    );

    const data = await response.json();

    assertEquals(response.status, 404);
    assertEquals(data, {
      error: {
        code: "error",
        message: "Feed item not found",
      },
    });
  });

  it("should return empty string if feed item does not provide a link", async () => {
    const feedItem = testFixtures.loadFeedItem(db, { link: null });

    using fetchSpy = spy(globalThis, "fetch");

    const response = await app.request(
      `/api/feed-items/${feedItem.id}/${feedItem.feedId}/extract-content`,
      {
        headers: {
          "authorization": `Basic ${encodeBase64("foo:bar")}`,
        },
      },
    );

    const data = await response.json();

    assertEquals(response.status, 200);
    assertEquals(data, { data: "" });
    assertSpyCalls(fetchSpy, 0);
  });

  it("should throw a 400 error if fetch call fails", async () => {
    const feedItem = testFixtures.loadFeedItem(db, {
      link: "https://example.com",
    });
    const fetchError = new Error("foo");

    using _ = stub(globalThis, "fetch", () => Promise.reject(fetchError));

    const response = await app.request(
      `/api/feed-items/${feedItem.id}/${feedItem.feedId}/extract-content`,
      {
        headers: {
          "authorization": `Basic ${encodeBase64("foo:bar")}`,
        },
      },
    );

    const data = await response.json();

    assertEquals(response.status, 400);
    assertEquals(data, {
      error: {
        code: "error",
        message: `Could not fetch "${feedItem.link}"`,
      },
    });
  });

  it("should throw a 400 error if fetch response is not ok", async () => {
    const feedItem = testFixtures.loadFeedItem(db, {
      link: "https://example.com",
    });

    using _ = stub(
      globalThis,
      "fetch",
      () => Promise.resolve(new Response("", { status: 400 })),
    );

    const response = await app.request(
      `/api/feed-items/${feedItem.id}/${feedItem.feedId}/extract-content`,
      {
        headers: {
          "authorization": `Basic ${encodeBase64("foo:bar")}`,
        },
      },
    );

    const data = await response.json();

    assertEquals(response.status, 400);
    assertEquals(data, {
      error: {
        code: "error",
        message: `Request to "${feedItem.link}" failed`,
      },
    });
  });

  it("should normalize page content links to open in new tab", async (t) => {
    const feedItem = testFixtures.loadFeedItem(db, {
      link: "https://example.com",
    });

    using fetchStub = stub(
      globalThis,
      "fetch",
      () =>
        Promise.resolve(
          new Response(
            `
          <html>
            <head></head>
            <body>
              <p>foo</p>
              <a href="">bar</a>
              <a href="/foo/bar">bar</a>
              <a href="https://example.com">bar</a>
              <a href="#foo">bar</a>
            </body>
          </html>
        `,
            { status: 200 },
          ),
        ),
    );

    const response = await app.request(
      `/api/feed-items/${feedItem.id}/${feedItem.feedId}/extract-content`,
      {
        headers: {
          "authorization": `Basic ${encodeBase64("foo:bar")}`,
        },
      },
    );

    const data = await response.json();

    assertEquals(response.status, 200);
    await assertSnapshot(t, data);
    assertSpyCalls(fetchStub, 1);
    assertSpyCallArg(fetchStub, 0, 0, feedItem.link);
  });

  it("should normalize page content links urls", async (t) => {
    const feedItem = testFixtures.loadFeedItem(db, {
      link: "https://example.com/a/b/c.html",
    });

    using fetchStub = stub(
      globalThis,
      "fetch",
      () =>
        Promise.resolve(
          new Response(
            `
          <html>
            <head></head>
            <body>
              <p>foo</p>
              <a href="/foo/bar">bar</a>
              <a href="./foo/bar">bar2</a>
              <a href="../foo/bar">bar3</a>
              <video src="/foo/bar" />
              <audio srcset="/foo/bar" />
              <a href="https://example.com">bar</a>
            </body>
          </html>
        `,
            { status: 200 },
          ),
        ),
    );

    const response = await app.request(
      `/api/feed-items/${feedItem.id}/${feedItem.feedId}/extract-content`,
      {
        headers: {
          "authorization": `Basic ${encodeBase64("foo:bar")}`,
        },
      },
    );

    const data = await response.json();

    assertEquals(response.status, 200);
    await assertSnapshot(t, data);
    assertSpyCalls(fetchStub, 1);
    assertSpyCallArg(fetchStub, 0, 0, feedItem.link);
  });

  it("should empty string if readablility cannot simplify content", async () => {
    const feedItem = testFixtures.loadFeedItem(db, {
      link: "https://example.com",
    });

    using _ = stub(
      globalThis,
      "fetch",
      () =>
        Promise.resolve(
          new Response(
            `
          <html>
            <head></head>
            <body></body>
          </html>
        `,
            { status: 200 },
          ),
        ),
    );

    const response = await app.request(
      `/api/feed-items/${feedItem.id}/${feedItem.feedId}/extract-content`,
      {
        headers: {
          "authorization": `Basic ${encodeBase64("foo:bar")}`,
        },
      },
    );

    const data = await response.json();

    assertEquals(response.status, 200);
    assertEquals(data, { data: "" });
  });

  it("should handle encoded uri component feed item id", async () => {
    const feedItem = testFixtures.loadFeedItem(db, {
      link: "https://example.com",
      id: "https://example.com",
    });

    using _ = stub(
      globalThis,
      "fetch",
      () =>
        Promise.resolve(
          new Response(
            `
          <html>
            <head></head>
            <body><p>oi</p></body>
          </html>
        `,
            { status: 200 },
          ),
        ),
    );

    const response = await app.request(
      `/api/feed-items/${
        encodeURIComponent(feedItem.id)
      }/${feedItem.feedId}/extract-content`,
      {
        headers: {
          "authorization": `Basic ${encodeBase64("foo:bar")}`,
        },
      },
    );

    const data = await response.json();

    assertEquals(response.status, 200);
    assertEquals(data, {
      data: '<div id="readability-page-1" class="page"><p>oi</p>\n' +
        "          \n" +
        "        </div>",
    });
  });
});
