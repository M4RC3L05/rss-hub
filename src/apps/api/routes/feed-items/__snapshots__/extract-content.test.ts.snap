export const snapshot = {};

snapshot[`global > GET /api/feed-items/:id/:feedId/extract-content > should normalize page content links to open in new tab 1`] = `
{
  data: '<div id="readability-page-1" class="page"><p>foo</p>
              <a href="" target="_blank">bar</a>
              <a href="https://example.com/foo/bar" target="_blank">bar</a>
              <a href="https://example.com/" target="_blank">bar</a>
              <a href="#foo">bar</a>
            
          
        </div>',
}
`;

snapshot[`global > GET /api/feed-items/:id/:feedId/extract-content > should normalize page content links urls 1`] = `
{
  data: '<div id="readability-page-1" class="page"><p>foo</p>
              <a href="https://example.com/foo/bar" target="_blank">bar</a>
              <a href="https://example.com/a/b/foo/bar" target="_blank">bar2</a>
              <a href="https://example.com/a/foo/bar" target="_blank">bar3</a>
              <video src="https://example.com/foo/bar">
              <audio srcset="https://example.com/foo/bar">
              <a href="https://example.com/" target="_blank">bar</a>
            
          
        </audio></video></div>',
}
`;
